EDX-29 - Пакетная обработка запросов апдейта статуса экзамена

dm-v opened the pull request.

dm-v committed:
EDX-29 - Пакетная обработка запросов апдейта статуса экзамена

dm-v committed:
EDX-29 - Пакетная обработка запросов апдейта статуса экзамена (обновление после ревью)

dm-v committed:
EDX-29 - Пакетная обработка запросов апдейта статуса экзамена (bugfix)

dm-v committed:
EDX-30 - Новый статус экзамена WA deleted_in_edx (+простановка версий для JS зависимостей)

dm-v committed:
EDX-33 - Заменить статус ok на ready_to_start при отображении

martynovp committed:
Comment event_start & event_finish fix

dm-v committed:
EDX-31 - Работа с комнатами. Не более одной активной комнаты единовременно.

dm-v committed:
EDX-34 - Работа с комнатами. Блокировка захода в экзамен.

dm-v committed:
EDX-36 - Убрать из WA хардкод адресов при обращении к sso/edx

dm-v committed:
Fix unittests

dm-v committed:
Migration to Python 3.5 + Django 2.0.3

martynovp commented on edx_proctor_webassistant/rest_framework.py:
```diff
@@ -0,0 +1,9 @@
+from rest_framework.pagination import PageNumberPagination
```
а это вообще зачем? почему явно не указывать 25/50?

martynovp replied:
в Django 2.0 так надо?

dm-v replied:
Это изменение относится к Django Rest Framework. Начиная с версии 3.3 нельзя больше использовать параметр `paginate_by`, нужно указывать `pagination_class`: http://www.django-rest-framework.org/topics/3.3-announcement/ А как переиспользовать дефолтный классов с педжинацией по 25/50 - я способа не нашел. В классе `PageNumberPagination` параметр `page_size` задан как `api_settings.PAGE_SIZE` - то есть задаётся где-то в настройках. Поэтому я решил объявить классы явно.

martynovp replied:
ок

martynovp commented on journaling/migrations/0002_auto_20160105_1350.py:
```diff
@@ -17,16 +17,16 @@ class Migration(migrations.Migration):
             model_name='journaling',
             name='event',
-            field=models.ForeignKey(blank=True, to='proctoring.EventSession', null=True),
+            field=models.ForeignKey(blank=True, to='proctoring.EventSession', null=True, on_delete=models.CASCADE),
```
а так вообще нормально? почему это не новая миграция?

dm-v replied:
Это нормально. в Django 2.0 нельзя использовать `ForeignKey` без параметра `on_delete` заданного явно. По-умолчанию это значение раньше задавалось как `CASCADE`. Если его не задать в старых миграциях, то сломаются тесты, т.к. тесты по умолчанию каждый раз нагоняют все миграции друг за другом на голую базу. На продакшене/QA данное изменение никак не отразится поскольку там эти миграции уже выполнены и повторно выполняться не будут

martynovp replied:
понял

martynovp commented on person/migrations/0001_initial.py:
```diff
@@ -19,7 +19,7 @@ class Migration(migrations.Migration):
                 ('object_type', models.CharField(max_length=64, choices=[(b'edxorg', 'Organization'), (b'edxcourse', 'Course'), (b'*', b'*'), (b'edxcourserun', 'Courserun')])),
                 ('role', models.CharField(default=b'proctor', max_length=10, choices=[(b'proctor', 'Proctor'), (b'instructor', 'Instructor')])),
-                ('user', models.ForeignKey(to=settings.AUTH_USER_MODEL)),
+                ('user', models.ForeignKey(to=settings.AUTH_USER_MODEL, on_delete=models.CASCADE)),
```
тот же вопрос, что и выше про миграции

dm-v replied:
ответил выше

martynovp replied:
ок

martynovp commented on proctoring/admin.py:
```diff
@@ -57,15 +75,102 @@ class EventSessionAdmin(admin.ModelAdmin):
+            } for exam in exams]
+
+            if codes:
+                max_retries = 3
```
логику объясни плз. почему и зачем три попытки сделаны

dm-v replied:
я такой трюк видал в паре мест в коде WA. например вот тут:
https://github.com/miptliot/edx_proctor_webassistant/blob/current_dev/proctoring/api_ui_views.py#L103
Сделано это для того, что если по какой-то причине запрос на изменение статусов не сработал (допустим сервер LMS перезагружался) - чтобы было еще пара попыток повторить запрос. Цифра 3 была взята чисто из головы. Конкретно в данном случае мы в админке реджектим все незавершенные аттемпты. Мне показалось, что в данной процедуре самое главное прибить сессию на WA, и если по какой-то причине не удалось отправить обновленные статусы в LMS - в принципе ничего страшного (сама процедура прибития сессии в WA не думается, что будет сильно частая). Как вариант могу переделать логику, и поместить все это дело в транзакцию - если запрос к LMS обламывется, я сбрасываю транзакцию и выдаю пользователю ошибку. надо?

martynovp replied:
нет, так нормально.

martynovp replied:
а не, стоп. если нам edx lms не вернул, что всё ок, нам надо сообщить об этом пользователю, а сессию тогда не прибивать, если у нас есть не финальные попытки. Что-нибудь типа
Ошибка при обращении к EDX. Пожалуйста, попробуйте ещё раз через одну минуту. При повторной ошибке, пожалуйста, обратитесь в службу поддержки <support_email>

dm-v replied:
хорошо, тогда я переведу управление транзакциями в мануальный режим и буду реджектить транзакцию в случае если edx вернул не OK

martynovp commented on proctoring/api_ui_views.py:
```diff
@@ -67,7 +68,7 @@ def get(self, request, attempt_code):
             exam.exam_status = exam.STARTED
             exam.proctor = request.user
-            exam.attempt_status = "OK"
+            exam.attempt_status = "ready_to_start"
```
а это у нас не в строковых константах?

dm-v replied:
Конкретно в WA везде по коду attempt-статусы задаются как строки. Вот в `edx-proctoring` - там да, там они используются как проперти какого-то там класса.
Конечно, в WA тоже можно создать некий константный класс, и приаттачить к нему проперти-статусы, но ИМХО это не сильно принципиально, можно и оставить так как есть.

martynovp replied:
Заведу задачу с low приоритетом в бэклог на это дело

martynovp commented on proctoring/api_ui_views.py:
```diff
@@ -209,49 +210,33 @@ def post(self, request):
+                    if exam and new_status and exam.attempt_status != new_status:
+                        if exam.attempt_status == 'ready_to_start' and new_status == 'started':
+                            exam.actual_start_date = datetime.now()
+                        if (exam.attempt_status == 'started'
```
на мой взгляд
`if new_status == 'submitted' and exam.attempt_status in ['started', 'ready_to_submit']`
более читаемо

dm-v replied:
поменяю - не вопрос

dm-v replied:
хм.... так это неправильно.
```
if (exam.attempt_status == 'started' and new_status == 'submitted') \
  or (exam.attempt_status == 'ready_to_submit' and new_status == 'submitted'):
```
вот это в 
```
if new_status == 'submitted' and exam.attempt_status in ['started', 'ready_to_submit']
```
трансформировать нельзя

martynovp commented on proctoring/edx_api.py:
```diff
@@ -61,12 +74,32 @@ def poll_status_request(codes):
 
+def poll_statuses_attempts(codes_list):
+    return requests.post(
+        settings.EDX_URL + "api/extended/edx_proctoring/proctoring_poll_statuses_attempts/",
```
надо будет как-нибудь везде "сложение" строк выпилить...

dm-v replied:
А что тебе сложение не нравится? оно человекочитаемое и понятное. Сложение через join дает минимальный прирост производительности. `+` если мне не изменяет память тормозил до версии 2.4 (или что-то вроде того). В современных версиях интерпретатора как ты складываешь строки на общую производительность вообще не влияет

martynovp replied:
я не про производительность, а про явное/неявное
В этом плане сложение требует либо явного приведения типов, либо необходимость быть полностью уверенным в отсутствии путаницы с типами. Тот же `format` или `%s` можно так сказать "уже обо всём за тебя подумал"

martynovp commented on proctoring/models.py:
```diff
@@ -9,6 +9,7 @@
 from django.utils.translation import ugettext_lazy as _
 from django.contrib.auth.models import User, AnonymousUser
 from django.db.models import Q
+from functools import reduce
```
???

dm-v replied:
http://python-future.org/compatible_idioms.html#reduce

dm-v replied:
с тройки `reduce` перенесли из глобальных функций в модуль `functools`

martynovp replied:
ок, понял

dm-v committed:
EDX-31 - Работа с комнатами. Не более одной активной комнаты единовременно (обновление после review)

dm-v committed:
EDX-35 - рефакторинг механизма нотификацией

dm-v committed:
EDX-35 - рефакторинг механизма нотификацией (bugfix)

dm-v committed:
EDX-35 - рефакторинг механизма нотификацией (bugfix)

dm-v committed:
EDX-59 - Убрать кнопку авторизации - сделать авторизацию автоматической сразу после открытия WA

martynovp committed:
EDX-66 Фикс работы проверки Auth.is_proctor()

martynovp committed:
EDX-67 Распознавать не указанный obj_type как любой (*)

martynovp committed:
EDX-68 Настройка назначения инструкторов прокторами

martynovp committed:
EDX-68 распознавание прав на конкретную сессию курса

dm-v committed:
EDX-74 - Создание бандлинга

dm-v committed:
EDX-60 - Переделка основного layout-а (header block, content block)

dm-v committed:
EDX-61 - Переделка страницы "Начать новую сессию" (новый API для "Последние созданные сессии" + переделка формы)
EDX-62 - Обновление библиотеки для grid-ов, приведение стиля grid-ов к виду как в мокапах

dm-v committed:
EDX-64 - Переделка основной страницы активной сессии (функционально добавляем только input для client side search)

dm-v committed:
EDX-63 - Переделка страницы "Архив сессий" (добавляем input для поиска)
EDX-65 - При просмотре сессии из архива должна открываться страница на подобии основной страницы активной сессии, но с ограниченными возможностями)
EDX-64 - Переделка основной страницы активной сессии (функционально добавляем только input для client side search) - bugfix

dm-v committed:
Translations

martynovp committed:
Merge branch 'new-design' into optimizations

mearsten committed:
Customizable DB port for Tornado App
